crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -3482,22 +3496,39 @@ impl<'db> TypeInferenceBuilder<'db> {
             ExprContext::Store => {
                 let value_ty = self.infer_expression(value);
 
-                let symbol = if let Type::Instance(instance) = value_ty {
-                    let instance_member = instance.class.instance_member(self.db(), attr);
-                    if instance_member.is_class_var() {
-                        self.context.report_lint(
+                let symbol = match value_ty {
+                    Type::Instance(instance) => {
+                        let instance_member = instance.class.instance_member(self.db(), attr);
+                        if instance_member.is_class_var() {
+                            self.context.report_lint(
                             &INVALID_ATTRIBUTE_ACCESS,
                             attribute.into(),
                             format_args!(
                                 "Cannot assign to ClassVar `{attr}` from an instance of type `{ty}`",
                                 ty = value_ty.display(self.db()),
                             ),
                         );
+                        }
+
+                        instance_member.0
                     }
+                    Type::ClassLiteral(_) | Type::SubclassOf(_) => {
+                        let class_member = value_ty.member(self.db(), attr);
 
-                    instance_member.0
-                } else {
-                    value_ty.member(self.db(), attr)
+                        if class_member.is_unbound() {
+                            self.context.report_lint(
+                            &INVALID_ATTRIBUTE_ACCESS,
+                            attribute.into(),
+                            format_args!(
+                                "Cannot assign to pure instance attribute `{attr}` from a class of type `{ty}`",
```

Comment:
I think we can omit the word "pure" here -- assigning to _any_ instance attribute from the class object is illegal

```suggestion
                                "Cannot assign to instance attribute `{attr}` from the class object `{ty}`",
```

Reply:
THe word "pure" comes from @sharkdp and afaik the intended target of this PR/issue. 

Do you mean that in red_knot you want to restrict this as well:

```python
class C:
    attr: str = "class_value"

    def __init__(self):
        self.attr = "instance value"

C.attr = "other class value"
```

if so, I think it is a deviation from what the issue said, so want to double-check that's the intent here.

Reply:
> Do you mean that in red_knot you want to restrict this as well:

Yes. We should allow instance attributes like that to be _accessed_ from the class object but not to be _assigned_ on the class object

Reply:
And for "pure" instance attributes -- those which do not have a default value set in the class body -- they should be neither readable nor writable on the class object :-)

Reply:
@AlexWaygood I applied naming changes per your suggestions. 

I did not implement the new case:

```python
class C:
    attr: str = "class_value"

    def __init__(self):
        self.attr = "instance value"

C.attr = "other class value"
```

for two reasons:

- outside of the scope of original issue
- more importantly: this would need changing member or onw_class_member API to return qualifiers. I think this will collide with what @sharkdp is doing with descriptors now. It would be much easier to add this one last case to tests & emit diagnostic when API's are changed

Reply:
> Yes. We should allow instance attributes like that to be _accessed_ from the class object but not to be _assigned_ on the class object. On the class object, they should be readable but not writable

What's the rationale for this? It's legal at runtime, both pyright and mypy allow it, I'm not sure I see why it must be disallowed.

Reply:
btw. I didn't argue with @AlexWaygood but I am with @carljm - IMHO there is no reason to forbid that. It was not and is not apparent to me

Reply:
> What's the rationale for this? It's legal at runtime, both pyright and mypy allow it, I'm not sure I see why it must be disallowed.

It's something I'd want a type checker to flag if I'm running a type checker on my Python code. Sure, it's legal at runtime, but it seems like it's probably a user mistake to me if the variable has been explicitly declared as an instance attribute but it's being modified on the class object rather than an instance. Having said all this, your pushback makes me think that the error should probably be disabled by default, since you're right that it isn't really _unsafe_ for an instance attribute with a class-level default to be assigned on the class object.

Reply:
As of current PR it will not emit a diagnostic. Do you suggest adding a configuration and implement it under a flag? Seems like a reasonable compromise. 

I haven't researched knot configuration capabilities yet. Maybe a follow up?

Reply:
@mishamsk you don't need to worry about it for now: it's not implemented on `main`, and it shouldn't be implemented as part of this PR :-) I was just responding to @carljm's (very reasonable) question about my earlier assertion. It could absolutely be a followup, but first we need to reach agreement on whether it's something we do indeed want ;-) and it might not be a priority for us to implement in any case